### DISCUSSION TITLE
Topgrade! Is there any way to determine whether a specific App uses Sparkle?

### DISCUSSION THREAD

**Discussion by kmplngj** [sparkle-project/Sparkle#2141] [Q&A]
Also, it doesn't seem that this cli can detect and upgrade all Sparkle enabled apps. Is there any way to determine whether a specific App uses Sparkle?

__Originally posted by @r-darwish in https://github.com/r-darwish/topgrade/issues/657#issuecomment-1126266720__

Topgrade is willing to integrate Spark Updates. Could someone perhaps help with some information in the mentioned issue? 

Thanks a lot!

**Comment by zorgiepoo** [comment]
Some heuristics for detecting if an app uses Sparkle are..
* Looking for `SUFeedURL`, `SUPublicEDKey`, or `SUPublicDSAKeyFile` (deprecated) keys in the app bundle's Info.plist (not perfect since old apps may have not migrated to using an EDKey yet and some apps may not specify the feed URL there)
* Looking for Sparkle.framework in the app bundle's Frameworks directory (not perfect if a developer forks Sparkle and renames it or places it in some nonstandard location for some reason)

If you instantiate a [SPUUpdater](https://sparkle-project.org/documentation/api-reference/Classes/SPUUpdater.html) targeting the `hostBundle` to the bundle you want to update, and if `-startUpdater:` returns an error, that means an external updater definitely cannot retrieve enough information (including the feed URL) to update the bundle.

Check how [sparkle-cli is implemented](https://github.com/sparkle-project/Sparkle/tree/2.x/sparkle-cli). It uses a [SPUUpdater](https://sparkle-project.org/documentation/api-reference/Classes/SPUUpdater.html) targeting the external bundle, and its own [SPUUserDriver](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUserDriver.html) for driving its own command line user interface.

Last time I checked, [Latest](https://github.com/mangerlahn/latest) for example drives updates using its own user driver, rather than using sparkle-cli. Both are possibilities, though.
